### PR TITLE
CAT-734: Adding get_auth_token() method

### PR DESCRIPTION
**Pull request by diontje**
@hohene @gpaulos 
Adding get_auth_token() method to fit_common

Test Run
(CAT-734_add_get_authentication_token_method) cuthbt1@socrates ~/sandbox/fit_tests/test/fit_tests $ python run_tests.py -stack 1 -test tests/bootstrap/torrey_function_test.py -v 3
get_auth_token test driver
==========================

get_auth_token() using global config
retrieving authentication token using credentials and port: {admin, admin123, port=8443}
my token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjoiYWRtaW4iLCJpYXQiOjE0NzIxNDk2NjAsImV4cCI6MTQ3MjIzNjA2MH0.Do1FYH5q8i92IC7yNyL5I_AtE4PmdsuaLOMXZohr57s

get_auth_token() using optional endpoint parameters (bad creds)
retrieving authentication token using credentials and port: {torrey1, torrey123, port=8443}
bad username or password, check GLOBAL_CONFIG or param endpoint values
Could not provide a token

get_auth_token() using optional endpoint parameters (good creds)
retrieving authentication token using credentials and port: {torrey, torrey123, port=8443}
my token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjoidG9ycmV5IiwiaWF0IjoxNDcyMTQ5NjYwLCJleHAiOjE0NzIyMzYwNjB9.tsKhOl_JrdsI-sLvlvXbppco6iHdfQQ6R2hXgzrY9dk
.
----------------------------------------------------------------------
Ran 1 test in 0.434s

OK

### CODE REVIEW

**Comment by gpaulos** [comment]
Thanks Torrey for the work.

At least one change needed. Jim says that the https port will be changed to 443. This is defined in GLOBAL_CONFIG['ports']['https']. Please retrieve from there.

Questions: 

How is the token used with API calls. Is it in the header? Can you give an example?

Was pycurl needed for this? Could this be done via requests? Trying to keep the number of entries in requirements.txt to minimum and have some continuity in the code form.

**Comment by diontje** [comment]
@gpaulos @hohene 

I use the token as we've done in the past.

Ex: 
curl -v -k -X DELETE https://stack1-ora.admin:8443/api/2.0/users/borg?auth_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjoidG9ycmV5IiwiaWF0IjoxNDcxOTg0MDIzLCJleHAiOjE0NzIwNzA0MjN9.uvDkgsUAJbj-cHV5vRT1ZtufUFXd5NLsy4CDucbfs2Y -d '{"username":"borg", "password":"borg123", "role":"Administrator" }' | python -m json.tool

However, seemingly the RackHD preferred way is documented in readthedocs or by simply using the 'TryItOut' button in the swagger-ui for each API

Ex:
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' --header 'Authorization: JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjoidG9ycmV5IiwiaWF0IjoxNDcyMTM2Njg4LCJleHAiOjE0NzIyMjMwODh9.6HhRv52cGojx6Qd408YH6wXIpFDcV9inxAXu23L5mOM' -d '{
  "username": "string",
  "password": "string",
  "role": "string"
}' 'http://stack1-ora.admin:8443/api/2.0/users'

PyCurl was chosen because of the need for testing security in the long run. Right now we are not currently producing signed OVAs but we do deliver them. This is piece of the build process that Dave was working going to deliver. I had a CAT story in the past to investigate signing OVAs via the OVFTool. I am not certain if he is responsible now.

Additionally, PyCurl gives you more control and options since its exactly the curl command we use everyday.  Here is an excerpt from the PR with the comment related to the current build process.

    # currently ignoring SSL certificate validation, ci-build ORA's currently do not have certs
    c.setopt(pycurl.SSL_VERIFYPEER, False)
    c.setopt(pycurl.SSL_VERIFYHOST, False)

    c.setopt(pycurl.HTTPHEADER, ['Content-Type: application/json'])
    c.setopt(pycurl.POST, 1)
    c.setopt(pycurl.CONNECTTIMEOUT, 5)
    c.setopt(pycurl.TIMEOUT, 8)
    c.setopt(pycurl.POSTFIELDS, login_payload)
    c.setopt(pycurl.WRITEDATA, result)

File: test/fit_tests/common/fit_common.py [hunk 1]
```diff
@@ -581,3 +584,71 @@ def _noserunner(pathspec):
...
+    :return: string = {'stdout': str:ouput, 'exitcode': return code}
+    '''
+    # override obtaining credentials from GLOBAL_CONFIG, rather provide endpoint data
+    if endpoint:
+        if {'username', 'password', 'port'} <= set(endpoint):
+            username = endpoint['username']
+            password = endpoint['password']
+            port = endpoint['port']
+        else:
+            return ''
```
Comment (hohene) [inline test/fit_tests/common/fit_common.py]: For these return failures, can you instead do something like this:
Initialize token = None  or token = ''  at start of funtion.
then return token
Easier on the user reading it.
token = get_auth_token(), if token is None:

**Comment by gpaulos** [comment]
Do you think we should migrate all API to pycurl instead of requests? Note that Onrack Conductor uses requests.

**Comment by diontje** [comment]
No. I do not.
1) Effectively PyCurl is simply curl
2) Yes the conductor uses requests, however, on-http and on-tftp, for example, uses urllib. The developer here could have chosen to use requests here was well presumably.
3)  Finally, this is open source. When a developer wants to contribute, their PR will not be denied because they do not use the preferred or familiar python module.  As long as it works and is readable, clean, efficient etc I think that is should be accepted. I am certainly not going to police someone's coding style.

**Comment by gpaulos** [comment]
OK. Can you put in that change I requested in comment above about the GLOBAL_CONFIG ports parameter and that suggestion from Erika? Then I will merge.

File: test/fit_tests/common/fit_common.py [hunk 2]
```diff
@@ -642,13 +643,13 @@ def get_auth_token(role=None, **endpoint):
...
+        return token
 
     # credential issue
     if c.getinfo(pycurl.HTTP_CODE) in [400, 401]:
         if VERBOSITY >= 3:
             print "bad username or password, check GLOBAL_CONFIG or param endpoint values"
-        return  ''
+        return token
 
     dic = json.loads(result.getvalue())
```
Comment (hohene) [inline test/fit_tests/common/fit_common.py]: what do you get if this fails or if dic ['token'] fails?
not sure if get works on a dictionary ....
token = dic.get('token', None)
return token

**Comment by diontje** [comment]
@hohene 
Done. You are right. I should have added an additional check there.

File: test/fit_tests/common/fit_common.py (hunk 2 above)
Comment (diontje) [inline test/fit_tests/common/fit_common.py]: @hohene 
Done. You are right. I should have added an additional check here.

**Comment by gpaulos** [comment]
+1

**Pull request merged**